flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/FatalProgramInvocationException.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.dispatcher;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.core.execution.ProgramInvocationException;
+
+import java.util.Collection;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Exception signalling that an exception occurred during the execution of the jar's main method.
+ *
+ * <p>The job will transition to FAILED state, and it will not be recovered.
+ */
+@Internal
+public class FatalProgramInvocationException extends ProgramInvocationException {
```

Comment:
Do we need a new exception type? As far as I can tell this does the same thing as ProgramInvocationException (we don't recover from that one either -> so it's also "fatal").

Reply:
Also PTAL at the `UnsuccessfulExecutionException`, that seems to be related.

Reply:
`Fatal` might not be the right word. WDYT about `HandledProgramInvocationException`?  I do not really want to make all `ProgramInvocationException` causing the program to go to a failed state.

I would see the normal `ProgramInvocationException` as a reason the JM shuts down and HA might recover and I want to introduce a new exception allowing a transition to a failed state.